### PR TITLE
graph-struct-1: graph structure concept

### PR DESCRIPTION
**Pull request by thequickf**
Usage example:
```C++
#include <graph.hpp>

class Node {
public:
    Node(int a, int b) : a(a), b(b) {}
    int a, b;
};

int main() {
    graph::Graph<int, graph::Net, graph::Weighted> g{{1, 2, 3}, {1, 3}};
    graph::Edge<Node> e{ {1, 2}, {1, 2} };
    g.addEdge({1, 2, 1});
    g.addEdge({2, 3, 2});
    g.addEdge({3, 1, 3});
    return 0;
}
```

### CODE REVIEW

File: src/graph.hpp [hunk 1]
```diff
@@ -2,15 +2,46 @@
...
+
+class EdgeTrait {
+protected:
+    EdgeTrait() = default;
+};
+
+// TODO: source and sink Node* type.
+// Should we inherit all nodes from base Node class
+// or should we apply more template woodoo?
+class Net : public GraphTrait {
```
Comment (isnullxbh) [inline src/graph.hpp:23]: Make `Net` a template class.

Comment (thequickf) [reply src/graph.hpp:23]: We have unpleasant interface for user in this case
```C++
graph::Graph<int, graph::Net<int>, graph::Weighted> g;
```
```graph::Net<int>``` user have to write ```int``` second time here (obviously it's unnecessary) and can make mistake

File: src/graph.hpp [hunk 2]
```diff
@@ -2,15 +2,46 @@
...
 };
 
+class Weighted : public EdgeTrait {
+public:
+    Weighted(uint32_t weight) : weight(weight) {}
+    uint32_t weight;
+};
+
+// TODO: Here I want to apply
+// requires ((std::is_base_of_v<graph::GraphTrait, Ts> || std::is_base_of_v<graph::EdgeTrait, Ts>) && ...)
```
Comment (isnullxbh) [inline src/graph.hpp]: Make a single trait that checks whether Ts is derived from GraphTrait or EdgeTrait:

```
template<typename T>
concept GraphTraitOrEdgeTrait =
    std::is_base_of_v<GraphTrait, T> or
    std::is_base_of_v<GraphTrait, T>;
```

and next:

`requires (GraphTraitOrEdgeTrait<Ts> && ...)`

File: src/impl/graph.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include <map>
+#include <set>
+#include <utility>
+
+namespace graph_impl {
+
+template<typename Node, typename GraphTraitList, typename EdgeTraitList>
+class Graph;
+
+template<
```
Comment (isnullxbh) [inline src/impl/graph.hpp:15]: OK, but in my view, this makes a code difficult to understand. I would prefer to pass template parameters for Edge separately.

File: src/graph.hpp (hunk 2 above)
Comment (thequickf) [reply src/graph.hpp]: Thanks!

File: src/graph.hpp (hunk 1 above)
Comment (isnullxbh) [reply src/graph.hpp:23]: It should result in a compile-time error. It is normal, people make mistakes.

File: src/impl/graph.hpp (hunk 3 above)
Comment (thequickf) [reply src/impl/graph.hpp:15]: The idea is it's hard to separate edge and graph traits from mathematics point of view.

Comment (isnullxbh) [reply src/impl/graph.hpp:15]: It is your choice, I just made my opinion :)

**Review by isnullxbh — APPROVED** [review]

File: src/graph.hpp (hunk 1 above)
Comment (thequickf) [reply src/graph.hpp:23]: Probably ```void*``` with ```reinterpret_cast``` will work here

Comment (isnullxbh) [reply src/graph.hpp:23]: It seems to me that is a bad idea to use void* in this case. You say:

> user have to write int second time here (obviously it's unnecessary) and can make mistake

, but at the same time suggest use `reinterpret_cast` with `void*`. Don't throw your code away, son. 😄

**Pull request merged**